tests/suite/graceful_recovery_test.go

```diff
@@ -98,8 +98,101 @@ var _ = Describe("Graceful Recovery test", Ordered, Label("functional", "gracefu
 	It("recovers when nginx container is restarted", func() {
 		runRecoveryTest(teaURL, coffeeURL, ngfPodName, nginxContainerName, files, &ns)
 	})
+
+	It("recovers when drained node is restarted", func() {
+		runRestartNodeWithDrainingTest(teaURL, coffeeURL, files, &ns)
+	})
+
+	It("recovers when node is restarted abruptly", func() {
+		runRestartNodeAbruptlyTest(teaURL, coffeeURL, files, &ns)
+	})
 })
 
+func runRestartNodeWithDrainingTest(teaURL, coffeeURL string, files []string, ns *core.Namespace) {
+	runRestartNodeTest(teaURL, coffeeURL, files, ns, true)
+}
+
+func runRestartNodeAbruptlyTest(teaURL, coffeeURL string, files []string, ns *core.Namespace) {
+	runRestartNodeTest(teaURL, coffeeURL, files, ns, false)
+}
+
+func runRestartNodeTest(teaURL, coffeeURL string, files []string, ns *core.Namespace, drain bool) {
+	nodeNames, err := getNodeNames()
+	Expect(err).ToNot(HaveOccurred())
+	Expect(nodeNames).To(HaveLen(1))
+
+	kindNodeName := nodeNames[0]
+
+	if portFwdPort != 0 {
+		close(portForwardStopCh)
+	}
+
+	if drain {
+		_, err := exec.Command(
+			"kubectl",
+			"drain",
+			kindNodeName,
+			"--ignore-daemonsets",
+			"--delete-local-data",
+		).CombinedOutput()
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = exec.Command("kubectl", "delete", "node", kindNodeName).CombinedOutput()
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	Expect(*clusterName).ToNot(BeEmpty())
+	containerName := *clusterName + "-control-plane"
+
+	_, err = exec.Command("docker", "restart", containerName).CombinedOutput()
+	Expect(err).ToNot(HaveOccurred())
+
+	// need to wait for docker container to restart and be running before polling for ready NGF Pods or else we will error
+	Eventually(
+		func() bool {
+			output, err := exec.Command(
+				"docker",
+				"inspect",
+				"-f",
+				"{{.State.Running}}",
+				containerName,
+			).CombinedOutput()
+			return strings.TrimSpace(string(output)) == "true" && err == nil
+		}).
+		WithTimeout(timeoutConfig.CreateTimeout).
+		WithPolling(500 * time.Millisecond).
+		Should(BeTrue())
+
+	// ngf can often oscillate between ready and error, so we wait for a stable readiness in ngf
```

Comment:
This is a solution I came up with since NGF would sometimes be ready, then error for a few seconds, then re-ready, then error, and it would take around ~40 seconds on a local Kind cluster for it to stabilize. I couldn't think of another solution that would successfully wait for it to stabilize. Does this seem alright?

Reply:
What was the error?

Reply:
Logs:

```
{"level":"info","ts":"2024-06-26T17:54:25Z","msg":"Starting NGINX Gateway Fabric in static mode","version":"edge","commit":"332aa29ab024c8ea49ac9621b28f4bae9e3b2ba8","date":"2024-06-26T17:14:48Z"}
failed to start control loop: error setting initial control plane configuration: NginxGateway nginx-gateway/my-release-config not found: failed to get API group resources: unable to retrieve the complete list of server APIs: gateway.nginx.org/v1alpha1: Get "https://10.96.0.1:443/apis/gateway.nginx.org/v1alpha1?timeout=10s": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

Describe events/information:
```
State:          Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Wed, 26 Jun 2024 10:54:25 -0700
      Finished:     Wed, 26 Jun 2024 10:54:35 -0700
    Last State:     Terminated
      Reason:       Unknown
      Exit Code:    255
      Started:      Wed, 26 Jun 2024 10:48:19 -0700
      Finished:     Wed, 26 Jun 2024 10:54:18 -0700
    Ready:          False

...
  Warning  Unhealthy       12s (x6 over 17s)  kubelet            Readiness probe failed: Get "http://10.244.0.5:8081/readyz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
  Warning  Unhealthy       11s                kubelet            Readiness probe failed: Get "http://10.244.0.5:8081/readyz": read tcp 10.244.0.1:48160->10.244.0.5:8081: read: connection reset by peer
  Warning  Unhealthy       11s                kubelet            Readiness probe failed: Get "http://10.244.0.5:8081/readyz": dial tcp 10.244.0.5:8081: connect: connection refused
  Warning  BackOff         11s                kubelet            Back-off restarting failed container nginx-gateway in pod my-release-nginx-gateway-fabric-54694cc9fc-rkzsd_nginx-gateway(d2f9e883-a0f7-485f-8ddf-2ecdebb3720b)
```

Reply:
Interesting. I think this is fine. I assume when the node restarts k8s has to recreate everything and the NginxGateway resource is not always created before NGF.